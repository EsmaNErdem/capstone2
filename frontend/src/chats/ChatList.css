.ChatList-Paper {
    width: 80%;
    height: 85vh;
    display: flex;
    justify-content: space-between;
    margin: auto;
}

.ChatList-Rooms {
    background-color: rgb(109, 23, 183, 0.12);
    overflow-y: auto;
}


.ChatList-Chat {
    height: 100%;
}


