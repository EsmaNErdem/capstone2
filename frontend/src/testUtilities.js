import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
