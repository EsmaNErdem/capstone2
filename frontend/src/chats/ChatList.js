import React, { useContext, useState, useEffect } from "react";
import UserContext from '../auth/UserContext';
import BookClubApi from "../api";
import Loading from "../utilities/Loading";
import Alert from "../utilities/Alert";
import ChatUser from "./ChatUser";
import { List, ListItem, ListItemText, Paper, Avatar } from "@mui/material";
import "./ChatList.css"

/**
 * Displays current user's previously open chat rooms
 * 
 * - Makes API call loads chat room data when the component mounts
 * - This component is designed for the "/chats" route.
 * - Utilizes Chat and Messages components and called by Routes
 * - Displays input text area for chat text input
 * - Lists previous messages between users
 * 
 * - Routes ==> ChatList ==> Chat ==> Messages
 */
const ChatList = () => {
    console.debug("ChatList");

    const { currentUser } = useContext(UserContext);
    const [rooms, setRooms] = useState([]);
    const [error, setError] = useState(null);
    const [receiver, setReceiver] = useState(null);
    const [loading, setLoading] = useState(true);
    
    /**
     * Fetches current user's chat rooms from backend
     */ 
    useEffect(() => {
        console.debug("Chatlist useEffect getUserChatRooms");
        
        // Make an API call to get chats room that were used by current user in chronological order
        const getRooms = async () => {
            setError(null);
            try {
                const userChatRooms = await BookClubApi.getUserPreviousMessages(currentUser.username);
                setRooms([...userChatRooms]);
                
                console.log(userChatRooms)
                if (Object.keys(userChatRooms).length !== 0) setReceiver(userChatRooms[0].user)
                else setError("No message yet")
            } catch (e) {
                console.error("ChatList useEffect API call data loading error:", e);
                setError("An error occurred while fetching chat rooms.");
            }
            
            setLoading(false)
        }

        // set loading to true while async getRooms runs; once the
        // data is fetched (or even if an error happens!), this will be set back
        // to true to control the spinner.
        setLoading(true)
        getRooms();
    }, []);

    /** Format message timestamp */ 
    const formatMessageDate = (messageDate) => {
        const options = { month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric', hour12: true };
        return new Date(messageDate).toLocaleString('en-US', options);
    };

    /* Handles click on room and start chat */
    const handleRoomClick = (username) => {
        setReceiver(username);
    };

    if (loading) return <Loading />;

    return (
        <div className="ChatList">
            {error ? <Alert type="danger" messages={[error]} />: null}
            
            <Paper elevation={6} className='ChatList-Paper' >
                <List className="ChatList-Rooms" style={{ flex: 2 }}>
                    {rooms.map((roomDetails) => (
                            <ListItem
                                key={roomDetails.room.id}
                                selected={receiver && receiver.username === roomDetails.user.username}
                                onClick={() => handleRoomClick(roomDetails.user)}
                            >
                                <Avatar alt={roomDetails.user.username} src={roomDetails.user.userImg} sx={{ width: 32, height: 32, marginRight: 2 }} />
                                <ListItemText 
                                    primary={roomDetails.user.username} 
                                    secondary={
                                        <React.Fragment>
                                            <span>{roomDetails.message.username === currentUser.username ? "You" : roomDetails.message.username}:</span>
                                            <span>{`${roomDetails.message.message} â€¢ ${formatMessageDate(roomDetails.message.messageDate)}`}</span>
                                        </React.Fragment>
                                    }
                                />
                            </ListItem> 
                    ))}
                </List>
                <List className="ChatList-Chat" style={{ flex: 4 }}>
                    {receiver && 
                        <ChatUser 
                            receiver={receiver.username} 
                            receiverImg={receiver.userImg}
                        />
                    }
                </List>

            </Paper>
        </div>
    );
    
}

export default ChatList;
