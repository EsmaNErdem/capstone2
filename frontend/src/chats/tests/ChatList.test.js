import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, act } from '@testing-library/react';
import ChatList from '../ChatList';
import BookClubApi from '../../api';

const mockChatRooms =  [{
    message: {
            username: 'TestUser',
            message: 'TestMessage', 
            messageDate: "02/07/2023",
            messageId: 1,
        },
    room: {
            id: 1,
            roomDate: "02/07/2023",
        },
    user: {
        username: 'TestUser',
        userImg: 'UserImg'
    }
},
{
    message: {
            username: 'TestUser2',
            message: 'TestMessage2', 
            messageDate: "02/07/2023",
            messageId: 2,
        },
    room: {
            id: 3,
            roomDate: "02/07/2023",
        },
    user: {
        username: 'TestUser2',
        userImg: 'UserImg2'
    }
}]

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        getUserPreviousMessages: jest.fn(),
    },
}));

afterEach(() => {
    jest.clearAllMocks();
});

  
test('renders without crashing', async () => {
    BookClubApi.getUserPreviousMessages.mockImplementationOnce(() => (mockChatRooms));
  
    await act(async () => {
      render(
        <MemoryRouter>
          <UserProvider>
            <ChatList />
          </UserProvider>
        </MemoryRouter>
      );
    });
  
});

test('it renders and matches with snapshot', async () => {
    BookClubApi.getUserPreviousMessages.mockImplementationOnce(() => (mockChatRooms));
  
    await act(async () => {
      ({ asFragment } = render(
        <MemoryRouter>
          <UserProvider>
            <ChatList />
          </UserProvider>
        </MemoryRouter>
      ));
    });
  
    expect(asFragment()).toMatchSnapshot();
});
  
test('renders mock data', async () => {
    BookClubApi.getUserPreviousMessages.mockImplementationOnce(() => (mockChatRooms));
  
    await act(async () => {
      ({ getByText, getAllByText } = render(
        <MemoryRouter>
          <UserProvider>
            <ChatList />
          </UserProvider>
        </MemoryRouter>
      ));
    });   
     const testUserElements = getAllByText("TestUser");
  
    expect(testUserElements.length).toBeGreaterThan(1);
    expect(getByText("TestMessage • Feb 7, 12:00 AM")).toBeInTheDocument()
    expect(getByText("TestUser2")).toBeInTheDocument()
    expect(getByText("TestMessage2 • Feb 7, 12:00 AM")).toBeInTheDocument()
});
