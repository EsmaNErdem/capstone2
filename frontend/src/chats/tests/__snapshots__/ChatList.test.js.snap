// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="ChatList"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 ChatList-Paper css-azc51k-MuiPaper-root"
    >
      <ul
        class="MuiList-root MuiList-padding ChatList-Rooms css-h4y409-MuiList-root"
        style="flex: 2;"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding Mui-selected css-1p823my-MuiListItem-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="TestUser"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg"
            />
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              TestUser
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              <span>
                TestUser:
              </span>
              <span>
                TestMessage • Feb 7, 12:00 AM
              </span>
            </p>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
          >
            <img
              alt="TestUser2"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="UserImg2"
            />
          </div>
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              TestUser2
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
            >
              <span>
                TestUser2:
              </span>
              <span>
                TestMessage2 • Feb 7, 12:00 AM
              </span>
            </p>
          </div>
        </li>
      </ul>
      <ul
        class="MuiList-root MuiList-padding ChatList-Chat css-h4y409-MuiList-root"
        style="flex: 4;"
      >
        <div
          class="ChatUser"
          style="width: 100%; height: 100%; display: flex; flex-direction: column; justify-content: space-between; padding-left: 5px;"
        >
          <a
            href="/profile/TestUser"
            style="text-decoration: none;"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular css-1lbvhy8-MuiAvatar-root"
              >
                <img
                  alt="TestUser"
                  class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="UserImg"
                />
              </div>
              <div
                class="MuiListItemText-root css-tlelie-MuiListItemText-root"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  style="color: orangered; font-size: 1.3rem;"
                >
                  TestUser
                </p>
              </div>
            </li>
          </a>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="Chat"
          >
            <div
              class="Chat-Messages"
            />
            <div
              class="Chat-ChatInput"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters ChatInput css-1nxmd3h-MuiListItem-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nlta2y-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                      id=":r1:"
                      rows="2"
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-ihdtdm"
                      >
                        <span
                          class="notranslate"
                        >
                          ​
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-11nzp8j-MuiButtonBase-root-MuiIconButton-root"
                  data-testid="chat-send-button"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="SendIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li
                class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
              />
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
